dm/pkg/shardddl/optimism/lock.go

```diff
@@ -871,3 +714,419 @@ func GetColumnName(lockID, ddl string, tp ast.AlterTableType) (string, error) {
 	}
 	return "", nil
 }
+
+// checkAddDropColumn check for ALTER TABLE ADD/DROP COLUMN statement
+// FOR ADD COLUMN, check whether add column with a different field or add a droped column
+// FOR DROP COLUMN, return the droped column.
+func (l *Lock) checkAddDropColumn(source, schema, table string, ddl string, prevTable, postTable schemacmp.Table) (string, error) {
+	currTable := l.tables[source][schema][table]
+	defer func() {
+		l.tables[source][schema][table] = currTable
+	}()
+
+	l.tables[source][schema][table] = prevTable
+	oldJoined, err := l.joinNormalTables()
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	l.tables[source][schema][table] = postTable
+	newJoined, err := l.joinNormalTables()
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	cmp, err := oldJoined.Compare(newJoined)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	if cmp <= 0 {
```

Comment:
dose cmp <=0 means it a drop column operateion ? seesm in this branch, you not return the dropped column when err not hanppened

Reply:
cmp <= 0 means a ADD column operation(oldJoin <= newJoin), so there is no dropped column. This used to check
```
source1: add column a varchar(4)
source2: add column a varchar(5)
```
then varchar(5) will be ignored(column already exist)